lib/pbench/server/api/resources/__init__.py

```diff
@@ -1567,27 +1597,88 @@ def _dispatch(
                 self._check_authorization(auth_params)
             except UnauthorizedAccess as e:
                 self.logger.warning("{}: {}", api_name, e)
+                if audit:
+                    Audit.create(
+                        root=audit,
+                        status=AuditStatus.FAILURE,
+                        reason=AuditReason.PERMISSION,
+                        attributes={"message": str(e)},
+                    )
                 abort(e.http_status, message=str(e))
             except Exception as e:
                 self.logger.exception("{}: {}", api_name, e)
+                if audit:
+                    Audit.create(
+                        root=audit,
+                        status=AuditStatus.FAILURE,
+                        reason=AuditReason.INTERNAL,
+                        attributes={"message": str(e)},
+                    )
                 abort(
                     HTTPStatus.INTERNAL_SERVER_ERROR,
                     message=HTTPStatus.INTERNAL_SERVER_ERROR.phrase,
                 )
 
+        # Pass the root audit object to the API implementation. Normally, we'll
+        # automatically generate an audit finalization; however if the API
+        # wants to emit a special audit sequence it can disable "finalize"
+        # in the context. It can also pass "attributes" by setting that
+        # field.
+        context = {
+            "auditing": {
+                "audit": audit,
+                "finalize": bool(audit),
+                "status": AuditStatus.SUCCESS,
+                "reason": None,
+                "attributes": None,
+            }
+        }
+
         try:
-            return execute(params, request)
+            response = execute(params, request, context)
+            auditing = context["auditing"]
+            if auditing["finalize"]:
+                Audit.create(
+                    root=auditing["audit"],
+                    status=auditing["status"],
+                    reason=auditing["reason"],
+                    attributes=auditing["attributes"],
+                )
+            return response
         except APIAbort as e:
-            self.logger.exception("{} {}", api_name, e)
+            auditing = context["auditing"]
+            if auditing["finalize"]:
+                attr = auditing.get("attributes", {"message": str(e)})
+                try:
+                    Audit.create(
+                        root=auditing["audit"],
+                        status=AuditStatus.FAILURE,
+                        reason=auditing["reason"],
+                        attributes=attr,
+                    )
```

Comment:
Yes, I would like to audit failures rather than leaving the unterminated `BEGIN`; otherwise we've got no idea what happened. The lifecycle of the resource was interrupted abnormally, and this is where we track details of the resource lifecycle. Partial failures in bulk operations are a special case, but I don't think that matters here.

Reply:
OK, leaving aside the partial failure case, I think logging failures just so that we won't have dangling `BEGIN` entries isn't an appropriate justification.  We can potentially address the dangling `BEGIN` entries in any of several ways:

- We can defer recording them until we get the successful `END`.
- We can record them initially and then delete them in the case of a failure (I don't mind putting extra junk on the failure path...).
- We can record them without committing the transaction and then roll it back in the event of a failure.

It's not obvious to me that having separate `BEGIN` and `END` entries is even useful.  For instance, it might be better if there were a _single_ entry for each transaction containing the begin- and end-information:  when the transaction is complete, we can either update or delete the entry depending on whether the transaction was successful or not.

Reply:
I don't want audit records to be edited. Begin and end audit records is a pretty widely established precedent, and I like the model because it brackets the operation *and* ensures that an abnormally terminated operation (like a crash of the server) still leaves a committed trail. I *have* considered generating an audit record but not committing it until later, though I'm not thrilled about that for several reasons.

Reply:
As a parting comment before I resolve this, I'll just offer that I think it would be best if the audit log tended toward the leaner, and, if we need more detail, we can look in the execution log.  In particular, I don't want either log (but, especially the audit log) to become a bottleneck.  So, doing less where we can seems like a good idea.